himbaechel/uarch/gatemate/route_clock.cc

```diff
@@ -107,125 +112,101 @@ void GateMateImpl::route_clock()
     }
 
     for (auto clk_net : clk_nets) {
-        log_info("    routing net '%s'\n", clk_net->name.c_str(ctx));
-        ctx->bindWire(ctx->getNetinfoSourceWire(clk_net), clk_net, STRENGTH_LOCKED);
-
-        auto clk_plane = 0;
-        switch (clk_net->driver.port.index) {
-        case id_GLB0.index:
-            clk_plane = 9;
-            break;
-        case id_GLB1.index:
-            clk_plane = 10;
-            break;
-        case id_GLB2.index:
-            clk_plane = 11;
-            break;
-        case id_GLB3.index:
-            clk_plane = 12;
-            break;
-        }
+        log_info("    routing net '%s' to %d users\n", clk_net->name.c_str(ctx), clk_net->users.entries());
+        auto src_wire = ctx->getNetinfoSourceWire(clk_net);
+        ctx->bindWire(src_wire, clk_net, STRENGTH_LOCKED);
 
+        auto sink_wires = dict<WireId, PortRef>{};
+        auto sink_wires_to_do = pool<WireId>{};
         for (auto &usr : clk_net->users) {
-            std::priority_queue<QueuedWire, std::vector<QueuedWire>, std::greater<QueuedWire>> visit;
-            dict<WireId, PipId> backtrace;
-            WireId dest = WireId();
-
             if (!feeds_clk_port(usr) && !feeds_ddr_port(clk_net, usr))
                 continue;
 
-            auto cpe_loc = ctx->getBelLocation(usr.cell->bel);
-            auto is_glb_clk = clk_net->driver.cell->type == id_GLBOUT;
-
             auto sink_wire = ctx->getNetinfoSinkWire(clk_net, usr, 0);
+
+            sink_wires.insert({sink_wire, usr});
+            sink_wires_to_do.insert(sink_wire);
+        }
+
+        std::priority_queue<QueuedWire, std::vector<QueuedWire>, std::greater<QueuedWire>> visit;
+        dict<WireId, PipId> backtrace;
+
+        auto is_glb_clk = clk_net->driver.cell->type == id_GLBOUT;
+
+        visit.push(QueuedWire(src_wire));
+        while (!visit.empty()) {
+            QueuedWire curr = visit.top();
+            visit.pop();
+            if (sink_wires_to_do.count(curr.wire)) {
+                if (ctx->debug) {
+                    auto sink_wire_name = "(uninitialized)";
+                    if (curr.wire != WireId())
+                        sink_wire_name = ctx->nameOfWire(curr.wire);
+                    log_info("            -> %s (%.3fns)\n", sink_wire_name, ctx->getDelayNS(curr.delay));
+                }
+                sink_wires_to_do.erase(curr.wire);
+                if (sink_wires_to_do.empty())
+                    break;
+            }
+
+            for (auto dh : ctx->getPipsDownhill(curr.wire)) {
+                if (!ctx->checkPipAvailForNet(dh, clk_net))
+                    continue;
+                WireId dst = ctx->getPipDstWire(dh);
+                if (backtrace.count(dst))
```

Comment:
this won't ever do revisits, just to check that this is intentional ?

Reply:
It's Dijkstra without lookahead, so each node is always visited with its lowest-cost path, aside from ties or whatever, so I think revisits aren't necessary.

Reply:
I don't think that's necessarily guaranteed. Imagine you have two ways of going between two wires:

```
    0.5ns         2.5ns
A -------> X -------> B
     1ns           0.5ns      0.5ns
A -------> Y ------> Z ------>  B
```

`X` will be visited first, and then the higher-cost path to `B`, the lower-cost path via `Y` and `Z` would not be discovered.

Note also the `alt < dist[v]` section in the "using a priority queue" pseudocode [here](https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm).